src/Common/tests/Collections/IEnumerableTest.cs

```diff
@@ -0,0 +1,428 @@
+                for (int i = 0; i < 3; i++)
+                {
+                    Assert.Throws<InvalidOperationException>(
+                        () => enumerator.Current);
+                }
+            }
+
+
+            int iterations;
+            if (CollectionOrder == CollectionOrder.Unspecified)
+            {
+                BitArray itemsVisited = new BitArray(needToMatchAllExpectedItems ? count : expectedItems.Length, false);
+                for (iterations = 0;
+                     iterations < count && enumerator.MoveNext();
+                     iterations++)
+                {
+                    var currentItem = enumerator.Current;
+                    var itemFound = false;
+                    for (int i = 0; i < itemsVisited.Length; ++i)
+                    {
+                        if (!itemsVisited[i]
+                            && Equals(currentItem, expectedItems[i + (needToMatchAllExpectedItems ? startIndex : 0)]))
+                        {
+                            itemsVisited[i] = true;
+                            itemFound = true;
+                            break;
+                        }
+                    }
+                    Assert.True(itemFound, "itemFound");
+
+                    for (int i = 0; i < 3; i++)
+                    {
+                        var tempItem = enumerator.Current;
+                        Assert.Equal(currentItem, tempItem);
+                    }
+                }
+                if (needToMatchAllExpectedItems)
+                {
+                    for (int i = 0; i < itemsVisited.Length; i++)
+                    {
+                        Assert.True(itemsVisited[i]);
+                    }
+                }
+                else
+                {
+                    var visitedItemCount = 0;
+                    for (int i = 0; i < itemsVisited.Length; i++)
+                    {
+                        if (itemsVisited[i])
+                            ++visitedItemCount;
+                    }
+                    Assert.Equal(count, visitedItemCount);
+                }
+            }
+            else
+            {
+                for (iterations = 0;
+                     iterations < count && enumerator.MoveNext();
+                     iterations++)
+                {
+                    var currentItem = enumerator.Current;
+                    Assert.Equal(expectedItems[iterations], currentItem);
+                    for (int i = 0; i < 3; i++)
+                    {
+                        var tempItem = enumerator.Current;
+                        Assert.Equal(currentItem, tempItem);
+                    }
+                }
+            }
+            Assert.Equal(count, iterations);
+
+            if (validateEnd)
+            {
+                for (int i = 0; i < 3; i++)
+                {
+                    Assert.False(
+                        enumerator.MoveNext(),
+                        "enumerator.MoveNext()");
```

Comment:
This is a free-form message to help identify the error, right?  "enumerator.MoveNext() returned true past the expected end", seems a bit more helpful.